File: go/webhook_test.go
package pinto

import (
	"bytes"
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestWebhookUtils(t *testing.T) {
	t.Run("VerifyWebhookSecret", func(t *testing.T) {
		if !VerifyWebhookSecret("secret123", "secret123") {
			t.Error("expected true on matching secret")
		}
		if VerifyWebhookSecret("wrong", "secret123") {
			t.Error("expected false on wrong secret")
		}
		if VerifyWebhookSecret("", "secret123") {
			t.Error("expected false on empty header")
		}
	})

	t.Run("ParseWebhookEvent", func(t *testing.T) {
		body := `{"event":"message.created","bot_id":"b1","chat_id":"c1","message":"hello","sender":{"user_id":"u1","name":"Bob"},"timestamp":"2026-09-06T12:00:00Z"}`
		req := httptest.NewRequest(http.MethodPost, "/webhook", bytes.NewBufferString(body))

		event, err := ParseWebhookEvent(req)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if event.Event != "message.created" {
			t.Errorf("wrong event: %s", event.Event)
		}
		if event.Sender.Name != "Bob" {
			t.Errorf("wrong sender: %s", event.Sender.Name)
		}
	})

	t.Run("NewReplyResponse", func(t *testing.T) {
		reply := NewReplyResponse("thanks", "https://img.jpg")
		if reply.ReplyMessage != "thanks" {
			t.Errorf("wrong reply message: %s", reply.ReplyMessage)
		}
		if reply.MediaURL != "https://img.jpg" {
			t.Errorf("wrong media url: %s", reply.MediaURL)
		}
	})
}

File: typescript/src/__tests__/webhook.test.ts
import { describe, expect, it } from 'vitest'
import {
  createReplyResponse,
  parseWebhookEvent,
  verifyWebhookSecret,
} from '../webhook.js'

describe('Webhook Utilities', () => {
  it('verifies webhook secret securely', () => {
    expect(verifyWebhookSecret('my-secret-key', 'my-secret-key')).toBe(true)
    expect(verifyWebhookSecret('wrong-secret', 'my-secret-key')).toBe(false)
    expect(verifyWebhookSecret(null, 'my-secret-key')).toBe(false)
    expect(verifyWebhookSecret('short', 'my-secret-key')).toBe(false)
  })

  it('parses incoming webhook message event', () => {
    const raw = JSON.stringify({
      event: 'message.created',
      bot_id: 'bot_123',
      chat_id: 'chat_456',
      message_id: 'msg_789',
      sender: { user_id: 'u_1', name: 'Alice' },
      message: 'Hello bot!',
      timestamp: '2026-09-06T12:00:00Z',
    })

    const event = parseWebhookEvent(raw)
    expect(event.event).toBe('message.created')
    if (event.event === 'message.created') {
      expect(event.sender.name).toBe('Alice')
      expect(event.message).toBe('Hello bot!')
    }
  })

  it('creates reply response payload correctly', () => {
    const reply = createReplyResponse('Hello from bot!', 'https://example.com/image.png')
    expect(reply.reply_message).toBe('Hello from bot!')
    expect(reply.media_url).toBe('https://example.com/image.png')
  })
})
